### PR TITLE
ARCH: COMMON: split sys_io.h for MMIO & memory bitops functions

### PR DESCRIPTION
**Pull request by evgeniy-paltsev**
As of today 'include/arch/common/sys_io.h" has generic implementation for MMIO accessors and memory bits manipulation functions. That leads to several architectures (like ARC, ARM/aarch64, ARM/aarch32/corter_a_r) redefine entire 'common/sys_io.h' even if they only have different MMIO accessors implementation.

Split 'include/arch/common/sys_io.h" to
 * sys_io.h - generic MMIO accessors implementation
 * sys_bitops.h - generic memory bits manipulation functions implementation

Use generic bitops implementation for:
 * ARC
 * ARM/aarch64
 * ARM/aarch32/corter_a_r

### CODE REVIEW

**Review by stephanosio — COMMENTED** [review]
Instead of completely getting rid of the arch-specific implementations, it would be better to have an option to specify whether arch-specific implementations are available/used (e.g. `CONFIG_ARCH_HAS_CUSTOM_BITOPS`).

File: include/arch/arm/aarch32/cortex_a_r/sys_io.h
Comment (andrewboie) [reply include/arch/arm/aarch32/cortex_a_r/sys_io.h:82]: In general we have always preferred to use builtins when available, we really want to purge as much of this arch-specific code from the kernel as we can. I would suggest that we not worry overmuch about less than optimal implementations in older GCC versions if new-ish versions are doing the right thing.

Comment (andrewboie) [reply include/arch/arm/aarch32/cortex_a_r/sys_io.h:82]: In other words: if the compiler isn't emitting optimal machine code for C like this, we have larger problems than just these particular functions, no?

**Review by tyrantlizard77 — APPROVED** [review]
A long time ago in 9df398b0d51d27df9e39cfab5291cea38fd66022 I added the sys_bitfield APIs. On ARC they were defined as C code you see here, on x86 they are #defined in arch.h:

```
+#define sys_bitfield_set_bit sys_set_bit
+#define sys_bitfield_clear_bit sys_clear_bit
+#define sys_bitfield_test_bit sys_test_bit
+#define sys_bitfield_test_and_set_bit sys_test_and_set_bit
+#define sys_bitfield_test_and_clear_bit sys_test_and_clear_bit
```

Later the ARC definitions (which were really just generic C code) were copied to include/arch/common/sys_io.h. AFAIK, everyone uses them....except x86.

The reason why x86 was defined like that (and in retrospect these really should have been wrapper inline functions, I think there was some kind of circular #include dependency issue at the time) was that x86 instructions like BTS if given a memory address treat it like a string. So the same functions work for both. https://www.felixcloutier.com/x86/bts

Thinking about it now, I'm really curious if GCC is smart enough to compile the generic defintions to use the same instructions. If so, I think we could at least move the sys_bitfield_* stuff out of arch/ and into a header under include/sys/ and have everything use it.

This is more of a side note. It's easy to move stuff if we decide to later. LGTM, please sort out the CI issues though. And let's get ACKs from all the different arch maintainers before we merge.

**Review by ruuddw — APPROVED** [review]
Makes sense to me to use the common C implementation, I expect compilers to be able to recognize the patterns and replace with the right instructions (probably even more optimal than the handwritten assembly variants, if used in larger expressions).

**Pull request merged**
